capsules/src/hmac.rs

```diff
@@ -246,26 +236,23 @@ impl<'a, H: digest::Digest<'a, T> + digest::HMACSha256, T: DigestType> digest::C
         });
     }
 
-    fn hash_done(&'a self, result: Result<(), ReturnCode>, digest: &'static mut T) {
+    fn hash_done(&'a self, result: Result<(), ErrorCode>, digest: &'static mut T) {
         self.appid.map(|id| {
             self.apps
                 .enter(*id, |app, _| {
                     self.hmac.clear_data();
 
                     let pointer = digest.as_ref()[0] as *mut u8;
 
-                    match app.dest.as_mut() {
-                        Some(dest) => {
-                            dest.as_mut().copy_from_slice(digest.as_ref());
-                        }
-                        None => {}
-                    };
-
-                    app.callback.map(|cb| match result {
-                        Ok(_) => cb.schedule(0, pointer as usize, 0),
-                        Err(e) => cb.schedule(usize::from(e), pointer as usize, 0),
+                    app.data.mut_map_or((), |dest| {
+                        dest.as_mut().copy_from_slice(digest.as_ref());
                     });
 
+                    match result {
+                        Ok(_) => app.callback.schedule(0, pointer as usize, 0),
+                        Err(e) => app.callback.schedule(usize::from(e), pointer as usize, 0),
```

Comment:
This looks like exactly the error I was hoping to prevent with the (as of now unresolved) discussion in #2320 . Because `ErrorCode` and `ReturnCode` use different values for the same errors, this now passes different error values to userspace than before. Which would require as-of-yet nonexistent changes in userspace to properly support. So for now, this should probably convert the error code to a return code, then to a usize, in order to not break any existing error handling code.